### PR TITLE
Extends Units module

### PR DESCRIPTION
**Pull request by cppcooper**
Adds unit check functions for
- visiting: ie. diplomats, merchants, or visitors
- visitor: ie. not merchants or diplomats
- invaders
- animals
- demons
- titans
- megabeasts
- semimegabeasts
- night creatures
- GreatDanger: true if unit is a demon, titan, or megabeast
- Danger: true if is a 'GreatDanger', semi-megabeast, night creature, undead, or invader

Units module: modifies existing checks
- ``isUndead(df::unit* unit)`` => ``isUndead(df::unit* unit, bool include_vamps = false)`` isUndead used to always ignore vamps, now it does it by default and includes them when true is passed
- ``isCitizen(df::unit* unit)`` => ``isCitizen(df::unit* unit, bool ignore_sanity = false)`` isCitizen used to always check sanity, now it does it by default and ignores sanity when true is passed

### CODE REVIEW

File: library/include/modules/Units.h [hunk 1]
```diff
@@ -115,7 +115,7 @@ DFHACK_EXPORT df::unit_misc_trait *getMiscTrait(df::unit *unit, df::misc_trait_t
 DFHACK_EXPORT bool isDead(df::unit *unit);
 DFHACK_EXPORT bool isAlive(df::unit *unit);
 DFHACK_EXPORT bool isSane(df::unit *unit);
-DFHACK_EXPORT bool isCitizen(df::unit *unit);
+DFHACK_EXPORT bool isCitizen(df::unit *unit, bool ignore_sanity = false);
```
Comment (myk002) [inline library/include/modules/Units.h:118]: how do the behavior of `isUndead` and `isCitizen` change when called from Lua? Does it respect the default parameter value? You might need to write a wrapper function that handles the optional params.

Comment (myk002) [reply library/include/modules/Units.h:118]: and update the Lua API docs

File: library/include/modules/Units.h [hunk 2]
```diff
@@ -138,9 +138,11 @@ DFHACK_EXPORT std::string getRaceBabyName(df::unit* unit);
 DFHACK_EXPORT std::string getRaceChildNameById(int32_t race_id);
 DFHACK_EXPORT std::string getRaceChildName(df::unit* unit);
 
+DFHACK_EXPORT bool isInvader(df::unit* unit);
```
Comment (myk002) [inline library/include/modules/Units.h:141]: Could you add the new functions to `dfhack_units_module` in LuaApi.cpp and document them in Lua API.rst?

**Review by myk002 — CHANGES_REQUESTED** [review]
nice!

File: library/include/modules/Units.h (hunk 2 above)
Comment (cppcooper) [reply library/include/modules/Units.h:141]: Dang, you mean they don't auto-export to lua

File: library/include/modules/Units.h (hunk 1 above)
Comment (cppcooper) [reply library/include/modules/Units.h:118]: Yea, it works as expected. I've tested via
```lua
[lua]# u = dfhack.gui.getSelectedUnit()
[lua]# print(u.mood)
-1
[lua]# print(dfhack.units.isCitizen(u))
true
[lua]# u.mood = 7
[lua]# print(dfhack.units.isCitizen(u))
false
[lua]# print(dfhack.units.isCitizen(u, false))
false
[lua]# print(dfhack.units.isCitizen(u, true))
true
```

Comment (cppcooper) [reply library/include/modules/Units.h:118]: > and update the Lua API docs

you'll have to let me know if what I've done is sufficient here

File: library/include/modules/Units.h (hunk 2 above)
Comment (cppcooper) [reply library/include/modules/Units.h:141]: exported.

File: docs/Lua API.rst [hunk 3]
```diff
@@ -1287,10 +1287,35 @@ Units module
 
   The unit is of the correct race of the fortress.
 
-* ``dfhack.units.isCitizen(unit)``
+* ``dfhack.units.isCitizen(unit[,ignore_sanity])``
 
   The unit is an alive sane citizen of the fortress; wraps the
-  same checks the game uses to decide game-over by extinction.
+  same checks the game uses to decide game-over by extinction
+  (except for the sanity check).
```
Comment (myk002) [inline docs/Lua API.rst:1294]:
```suggestion
  same checks the game uses to decide game-over by extinction.
  You can identify citizens, regardless of their sanity, by passing ``true``
  as the optional second parameter.
```

File: docs/Lua API.rst [hunk 4]
```diff
@@ -1287,10 +1287,35 @@ Units module
...
+
+  The unit is either a merchant, diplomat, or plain visitor.
+
+* ``dfhack.units.isVisitor(unit)``
+
+  The unit is strictly a visitor. Merchants and diplomats do not count here.
+
+* ``dfhack.units.isUndead(unit[,include_vamps])``
+
+  The unit is undead, but not a vampire.
```
Comment (myk002) [inline docs/Lua API.rst:1310]:
```suggestion
  The unit is undead. Pass ``true`` as the optional second parameter to
  count vampires as undead.
```

File: library/LuaApi.cpp [hunk 5]
```diff
@@ -1615,9 +1615,11 @@ static const LuaWrapper::FunctionReg dfhack_units_module[] = {
     WRAPM(Units, getRaceBabyNameById),
     WRAPM(Units, getRaceChildName),
     WRAPM(Units, getRaceChildNameById),
+    WRAPM(Units, isInvader),
     WRAPM(Units, isBaby),
     WRAPM(Units, isChild),
     WRAPM(Units, isAdult),
+    WRAPM(Units, isAnimal),
```
Comment (myk002) [inline library/LuaApi.cpp:1622]: needs Lua docs

File: library/LuaApi.cpp [hunk 6]
```diff
@@ -1640,6 +1644,13 @@ static const LuaWrapper::FunctionReg dfhack_units_module[] = {
     WRAPM(Units, isKilled),
     WRAPM(Units, isGelded),
     WRAPM(Units, isDomesticated),
+    WRAPM(Units, isDemon),
+    WRAPM(Units, isTitan),
+    WRAPM(Units, isMegabeast),
+    WRAPM(Units, isGreatDanger),
+    WRAPM(Units, isSemiMegabeast),
+    WRAPM(Units, isNightCreature),
+    WRAPM(Units, isDanger),
```
Comment (myk002) [inline library/LuaApi.cpp:1653]: needs docs for isDemon, isTitan, isMegabeast, isSemiMegabeast, and isNightCreature

**Review by myk002 — CHANGES_REQUESTED** [review]

File: docs/Lua API.rst (hunk 3 above)
Comment (cppcooper) [reply docs/Lua API.rst:1294]: Why drop the information about the sanity check? Shouldn't they know? Especially if the doc makes a point of talking about the closeness to the game's check

Comment (myk002) [reply docs/Lua API.rst:1294]: What did I drop? I thought I just reworded the (except for the sanity check) to make it clearer that you could choose to deviate from the default behavior by passing an optional parameter.

Comment (cppcooper) [reply docs/Lua API.rst:1294]: It just says it is the same as the game's now. The sanity stuff is an entirely separate statement that can't be inferred as related to the game's default behaviour.

Comment (myk002) [reply docs/Lua API.rst:1294]: I thought the game's default behavior was to not count insane dwarves for the purposes of game over. This flag allows you to modify the default behavior so you can count insane dwarves as citizens

Comment (cppcooper) [reply docs/Lua API.rst]: According to the function comments, this isn't the case. The game considers the insane to be citizens. It explicitly mentioned it, which is why I did.

Comment (cppcooper) [reply docs/Lua API.rst]:
```
    // Copied from the conditions used to decide game over,
    // except that the game appears to let melancholy/raving
    // dwarves count as citizens.
```

Comment (cppcooper) [reply docs/Lua API.rst]: though, maybe the berserk/crazed check (part of isSane) is part of the game's check??

File: docs/Lua API.rst [hunk 7]
```diff
@@ -1188,6 +1188,244 @@ Job module
...
+
+* ``dfhack.units.isAdult(unit)``
+
+  The unit is an adult.
+
+* ``dfhack.units.isGay(unit)``
+
+  The unit is gay.
+
+* ``dfhack.units.isNake(unit)``
```
Comment (cppcooper) [inline docs/Lua API.rst:1288]:
```suggestion
* ``dfhack.units.isNaked(unit)``
```

File: docs/Lua API.rst [hunk 8]
```diff
@@ -1188,6 +1188,244 @@ Job module
...
+
+  The unit is a child.
+
+* ``dfhack.units.isAdult(unit)``
+
+  The unit is an adult.
+
+* ``dfhack.units.isGay(unit)``
+
+  The unit is gay.
```
Comment (cppcooper) [inline docs/Lua API.rst:1286]: Maybe a lot of these could just drop the description part of the equation.. Most of these are pretty self-explanatory via their function names

File: docs/Lua API.rst [hunk 9]
```diff
@@ -1188,6 +1188,244 @@ Job module
...
+
+* ``dfhack.units.isNaked(unit)``
+
+  The unit is naked.
+
+* ``dfhack.units.isVisiting(unit)``
+
+  The unit is visiting. eg. Merchants, Diplomatics, travelers.
+
+
```
Comment (myk002) [inline docs/Lua API.rst:1296]:
```suggestion

```

File: docs/Lua API.rst (hunk 8 above)
Comment (myk002) [reply docs/Lua API.rst:1286]: yeah, feel free to group these simple functions with a common generic description string (e.g. "Checks simple attributes of the unit")

File: docs/Lua API.rst (hunk 9 above)
Comment (myk002) [reply docs/Lua API.rst:1296]: unless these extra lines (here an elsewhere in this PR) are here for readability, I'd say get rid of them to keep the document consistent

File: docs/Lua API.rst [hunk 10]
```diff
@@ -1188,6 +1188,244 @@ Job module
 Units module
 ------------
 
+* ``dfhack.units.isUnitInBox(unit,x1,y1,z1,x2,y2,z2)``
```
Comment (myk002) [inline docs/Lua API.rst:1191]: getUnitsInBox also needs documenting, I think

File: library/include/modules/Units.h [hunk 11]
```diff
@@ -68,6 +68,73 @@ static const int MAX_COLORS = 15;
  * The Units module - allows reading all non-vermin units and their properties
  */
 
+DFHACK_EXPORT bool isUnitInBox(df::unit* u,
+int16_t x1, int16_t y1, int16_t z1,
+int16_t x2, int16_t y2, int16_t z2);
```
Comment (myk002) [inline library/include/modules/Units.h:73]: needs formatting (indenting? keep on one long line like some other functions in this file?

File: library/include/modules/Units.h [hunk 12]
```diff
@@ -99,34 +166,11 @@ DFHACK_EXPORT df::language_name *getVisibleName(df::unit *unit);
...
-DFHACK_EXPORT bool isWar(df::unit* unit);
-DFHACK_EXPORT bool isHunter(df::unit* unit);
-DFHACK_EXPORT bool isAvailableForAdoption(df::unit* unit);
-DFHACK_EXPORT bool isOwnCiv(df::unit* unit);
-DFHACK_EXPORT bool isOwnGroup(df::unit* unit);
-DFHACK_EXPORT bool isOwnRace(df::unit* unit);
-DFHACK_EXPORT bool isVisible(df::unit* unit);
-/// is unit hidden to the player? accounts for ambushing
-DFHACK_EXPORT bool isHidden(df::unit *unit);
+        DFHACK_EXPORT df::unit_misc_trait *getMiscTrait(df::unit *unit, df::misc_trait_type type, bool create = false);
```
Comment (myk002) [inline library/include/modules/Units.h:173]: remove rogue indent

**Review by myk002 — CHANGES_REQUESTED** [review]

File: docs/Lua API.rst (hunk 9 above)
Comment (cppcooper) [reply docs/Lua API.rst:1296]: They are there for readability, they can parsed out. But they match the header's spacing

File: library/include/modules/Units.h (hunk 12 above)
Comment (cppcooper) [reply library/include/modules/Units.h:173]: counter: indent all of the declarations?

File: library/include/modules/Units.h (hunk 11 above)
Comment (cppcooper) [reply library/include/modules/Units.h:73]: weird, dunno how it lost its indentation

File: docs/Lua API.rst (hunk 10 above)
Comment (cppcooper) [reply docs/Lua API.rst:1191]: It is, ctrl+f
I put all the is* functions at the top. I think get in box is the first one under

File: library/include/modules/Units.h (hunk 12 above)
Comment (myk002) [reply library/include/modules/Units.h]: either is good

**Review by myk002 — APPROVED** [review]

File: library/include/modules/Units.h (hunk 12 above)
Comment (cppcooper) [reply library/include/modules/Units.h]: I guess I'll keep this in mind for the next one lol

**Pull request merged**